quantmsutils/mzml/mzml_statistics.py

```diff
@@ -1,20 +1,157 @@
 import re
 import sqlite3
 from pathlib import Path
+from typing import Optional
 
 import click
+import numpy as np
 import pandas as pd
-import pyarrow
-from pyopenms import MSExperiment, MzMLFile
+import pyarrow as pa
+import pyarrow.parquet as pq
+from pyopenms import MzMLFile
 
 
+class BatchWritingConsumer:
+    def __init__(self, parquet_schema, output_path, batch_size=10000, id_only=False):
+        self.parquet_schema = parquet_schema
+        self.output_path = output_path
+        self.batch_size = batch_size
+        self.id_only = id_only
+        self.batch_data = []
+        self.psm_parts = []
+        self.parquet_writer = None
+        self.acquisition_datetime = None
+        self.scan_pattern = re.compile(r"[scan|spectrum]=(\d+)")
+
+    def setExperimentalSettings(self, settings):
+        # Extract acquisition date/time from experimental settings
+        self.acquisition_datetime = settings.getDateTime().get()
+
+    def setExpectedSize(self, a, b):
+        pass
+
+    def consumeChromatogram(self, chromatogram):
+        # Optionally handle chromatograms if needed
+        pass
+
+    def consumeSpectrum(self, spectrum):
+        """
+        Consume spectrum data and write to parquet file.
+        :param spectrum: spectrum data
+        :return:
+        """
+
+        peaks = spectrum.get_peaks()
+        mz_array, intensity_array = peaks[0], peaks[1]
+        peak_per_ms = len(mz_array)
+        base_peak_intensity = float(np.max(intensity_array)) if peak_per_ms > 0 else None
+        total_intensity = float(np.sum(intensity_array)) if peak_per_ms > 0 else None
+        ms_level = spectrum.getMSLevel()
+        rt = spectrum.getRT()
+
+        if ms_level == 2:
+            precursor = spectrum.getPrecursors()[0]
+            charge_state = precursor.getCharge()
+            exp_mz = precursor.getMZ()
+
+            if self.id_only:
+                scan_id = self.scan_pattern.findall(spectrum.getNativeID())[0]
+                self.psm_parts.append([
+                    scan_id, ms_level,
+                    mz_array.tolist(),
```

Comment:
This and the next line is very wasteful.
You are copying large vectors into a python list, just to get it into a C array for pyarrow again later.
Try to keep it as numpy arrays.

Reply:
Let me double-check that one, I was more focused on the previous logic, not on the id_only part. I will improve it now.

Reply:
Just remove the two to list calls here. Maybe it works out of the box.

Reply:
@jpfeuffer Let me double-check: the way to write into parquet in batches is using a batch table, but the only way I have to create a table is with lists. I will double-check what I can do.

Reply:
@jpfeuffer I was reading more about it and it looks like using `pa.RecordBatch.from_pylist` is better than converting the entire table. https://arrow.apache.org/docs/python/generated/pyarrow.RecordBatch.html#

Reply:
I never said anything about from_pylist...

Just remove the .tolist please